dialogue-serde/src/main/java/com/palantir/conjure/java/dialogue/serde/DeferredEncoding.java

```diff
@@ -0,0 +1,94 @@
+/*
+ * (c) Copyright 2020 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.conjure.java.dialogue.serde;
+
+import com.google.common.base.Suppliers;
+import com.palantir.dialogue.TypeMarker;
+import com.palantir.logsafe.Preconditions;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.function.Supplier;
+
+/**
+ * Wrapper around an {@link Encoding} which allows both {@link Serializer} and {@link Deserializer} instances
+ * to be created when they're first used. This avoids the up-front cost of creating jackson <pre>ObjectReader</pre>
+ * and <pre>ObjectWriter</pre> instances for each request and response body type, for each encoding on service
+ * startup when many endpoints are only used with one encoding based on the clients that make requests.
+ * Note that this results in the first request to a given endpoint being more expensive than it would be
+ * otherwise, though this is already the case to an extent before the JIT compiler can optimize the path.
+ */
+final class DeferredEncoding implements Encoding {
+
+    private final Encoding delegate;
+
+    DeferredEncoding(Encoding delegate) {
+        this.delegate = Preconditions.checkNotNull(delegate, "Encoding is required");
+    }
+
+    @Override
+    public <T> Serializer<T> serializer(TypeMarker<T> type) {
+        return new DeferredSerializer<>(() -> delegate.serializer(type));
+    }
+
+    @Override
+    public <T> Deserializer<T> deserializer(TypeMarker<T> type) {
+        return new DeferredDeserializer<>(() -> delegate.deserializer(type));
+    }
+
+    @Override
+    public String getContentType() {
+        return delegate.getContentType();
+    }
+
+    @Override
+    public boolean supportsContentType(String contentType) {
+        return delegate.supportsContentType(contentType);
+    }
+
+    @Override
+    public String toString() {
+        return "DeferredEncoding{delegate=" + delegate + '}';
+    }
+
+    private static final class DeferredSerializer<T> implements Serializer<T> {
+
+        private final Supplier<Serializer<T>> delegate;
+
+        DeferredSerializer(Supplier<Serializer<T>> delegate) {
+            this.delegate = Suppliers.memoize(delegate::get);
```

Comment:
given that these creation functions have no side-effects, thoughts on just storing a nullable field and not using Suppliers.memoize with it's `synchronized` block?

Reply:
memoize doesn't synchronize unless another thread is already creating the object, in that case I think the ObjectReader/ObjectWriter are expensive enough that we might as well block and reuse the instance another thread has created for us. After the first creation, the memoized supplier is a simple volatile read.